definitions/npm/firebase_v4.x.x/flow_v0.46.x-/firebase_v4.x.x.js

```diff
@@ -0,0 +1,410 @@
+  delete(): Promise<void>;
+  getIdToken(forceRefresh?: boolean): Promise<string>;
+  getToken(forceRefresh?: boolean): Promise<string>;
+  linkAndRetrieveDataWithCredential(credential: AuthCredential): Promise<UserCredential>;
+  linkWithCredential(credential: AuthCredential): Promise<User>;
+  linkWithPhoneNumber(phoneNumber: string, applicationVerifier: ApplicationVerifier): Promise<ConfirmationResult>;
+  linkWithPopup(provider: OAuthProvider): Promise<UserCredential>;
+  reauthenticateAndRetrieveDataWithCredential(credential: AuthCredential): Promise<void>;
+  reauthenticateWithCredential(credential: AuthCredential): Promise<void>;
+  reauthenticateWithPhoneNumber(phoneNumber: string, applicationVerifier: ApplicationVerifier): Promise<ConfirmationResult>;
+  reload(): Promise<void>;
+  sendEmailVerification(): Promise<void>;
+  toJSON(): Object;
+  unlink(providerId: string): Promise<User>;
+  updateEmail(newEmail: string): Promise<void>;
+  updatePassword(newPassword: string): Promise<void>;
+  updatePhoneNumber(phoneCredential: AuthCredential): Promise<void>;
+  updateProfile(profile: { displayName: ?string, photoURL: ?string }): Promise<void>;
+}
+
+declare class EmailAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  providerId: string;
+  credential(email: string, password: string): AuthCredential;
+}
+
+declare class FacebookAuthProvider extends AuthProvider  {
+  PROVIDER_ID: string;
+  credential(token: string): AuthCredential;
+  addScope(scope: string): FacebookAuthProvider;
+  setCustomParameters(customOAuthParameters: Object): FacebookAuthProvider;
+}
+
+declare class GithubAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  credential(token: string): AuthCredential;
+  addScope(scope: string): GithubAuthProvider;
+  setCustomParameters(customOAuthParameters: Object): GithubAuthProvider;
+}
+
+declare class GoogleAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  credential(idToken?: string, accessToken?: string): AuthCredential;
+  addScope(scope: string): GoogleAuthProvider;
+  setCustomParameters(customOAuthParameters: Object): GoogleAuthProvider;
+}
+
+declare class PhoneAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  constructor(auth?: Auth): PhoneAuthProvider;
+  credential(verificationId: string, verificationCode: string): AuthCredential;
+  verifyPhoneNumber(phoneNumber: string, applicationVerifier: ApplicationVerifier): Promise<string>;
+}
+
+declare class TwitterAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  credential(token: string, secret: string): AuthCredential;
+  setCustomParameters(customOAuthParameters: Object): this;
+}
+
+declare type OAuthProvider = | FacebookAuthProvider
+                             | GithubAuthProvider
+                             | GoogleAuthProvider
+                             | TwitterAuthProvider;
+
+/****** database ******/
+declare type FirebaseValue = {} | Array<FirebaseValue> | string | number | boolean | null;
```

Comment:
Thanks for responding back to me! The code below is a simple case of what I have trouble with. Which is why I started looking into how what typescript did for their types. 

```
/* @flow */
type FirebaseValue = { [string| number]: FirebaseValue } | Array<FirebaseValue> | string | number | boolean | null;
type Workbook = { connection: { label: string } };

function update(x: { [string | number] : FirebaseValue }) {
  console.log(x);
}

const wb: Workbook = { connection: { label: 'test' }};

update(wb);
```

Another case is where I'm unsure what I need to do to be able to store my value into the workbook value

```
/* @flow */
type FirebaseValue = { [string| number]: FirebaseValue } | Array<FirebaseValue> | string | number | boolean | null;

class Test {
  workbook: { connection: { label: string } };
  snap = (snapshot: FirebaseValue) => {
    if (snapshot instanceof Object) {
      if (snapshot.connection && snapshot.connection instanceof Object) {
        this.workbook = snapshot;
      }
    }
  }
}
```